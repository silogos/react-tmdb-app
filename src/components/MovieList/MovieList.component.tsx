import { DisplayError, DisplayNoData } from "@/components/Displays";

function MovieCardShimmenring() {
  return (
    <div className="animate-pulse aspect-[8/12] rounded-lg bg-gray-700 bg-opacity-90" />
  );
}

export function ListLoading() {
  return Array(20)
    .fill(true)
    .map((_, idx) => <MovieCardShimmenring key={idx.toString()} />);
}

export function ListError() {
  return <DisplayError />;
}

export function ListEmpty() {
  return <DisplayNoData />;
}
