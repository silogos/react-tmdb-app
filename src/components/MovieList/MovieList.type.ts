import { Movie } from "@/types/Movie.types";

export type MovieListProps = {
  movies: Movie[];
  isLoading: boolean;
  isError: boolean;
  onReload?: () => void;
  hasMoreData: boolean;
  onLoadMore: () => void;
};

export type MoviListErrorProps = {
  onReload?: () => void;
};
