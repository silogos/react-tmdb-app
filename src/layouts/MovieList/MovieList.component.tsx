export function Hero() {
  return (
    <section id="hero" className="overflow-hidden py-4">
      <div className="mx-auto max-w-screen-xl px-4 py-14 lg:px-8 lg:py-20">
        <h1 className="text-center text-gray-300 text-4xl lg:text-7xl mb-4">
          ðŸŽ¬ Movie App
        </h1>
        <p className="text-center text-gray-300 text-base lg:text-2xl">
          (The Movie Database)
        </p>
      </div>
    </section>
  );
}
