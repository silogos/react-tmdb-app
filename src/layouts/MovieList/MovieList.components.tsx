import { useNavigate } from "react-router";

export function MovieListHero() {
  return (
    <section id="hero" className="overflow-hidden py-4">
      <div className="mx-auto max-w-screen-xl px-4 py-14 lg:px-8 lg:py-20">
        <h1 className="text-center text-gray-300 text-4xl lg:text-7xl mb-4">
          TMDB
        </h1>
        <p className="text-center text-gray-300 text-base lg:text-2xl">
          (The Movie Database)
        </p>
      </div>
    </section>
  );
}

export function SearchInput() {
  const navigate = useNavigate();

  return (
    <form
      onSubmit={(evt) => {
        evt.preventDefault();
        console.log("Submited");
      }}
    >
      <input
        type="search"
        className="w-64 border-2 rounded-lg  px-4 py-2 bg-gray-300"
        placeholder="Search..."
        onKeyDown={(evt) => {
          if (evt.currentTarget.checkValidity() && evt.key === "Enter") {
            console.log("do validate", !!evt.currentTarget.value);
            navigate(`/movies/search?query=${evt.currentTarget.value}`);
          }
        }}
        minLength={3}
        required
      />
    </form>
  );
}
