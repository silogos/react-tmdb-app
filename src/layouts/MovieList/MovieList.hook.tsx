import { useNavigate } from "react-router";
import { useCallback } from "react";

function useMovieListLayout() {
  const navigate = useNavigate();
  const handleSearch = useCallback((searchText: string) => {
    navigate(`/movies/search?query=${searchText}`);
  }, []);

  return {
    handleSearch,
  };
}

export default useMovieListLayout;
