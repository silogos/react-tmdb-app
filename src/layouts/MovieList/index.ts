export { default } from "./MovieList.layout";
