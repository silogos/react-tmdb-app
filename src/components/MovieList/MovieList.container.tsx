import MovieCard from "@/components/MovieCard";
import { MovieListProps } from "./MovieList.type";
import { ListEmpty, ListError, ListLoading } from "./MovieList.component";

function MovieList({
  movies,
  isLoading,
  isError,
  hasMoreData,
  onLoadMore,
}: MovieListProps) {
  return (
    <>
      <div className="grid grid-cols-2 md:grid-cols-4 lg:grid-cols-5 gap-y-8 gap-4">
        {movies.map((movie) => (
          <MovieCard
            key={movie.id}
            id={movie.id}
            poster={movie.poster_path}
            releaseDate={movie.release_date}
            title={movie.title}
            voteAverage={movie.vote_average}
          />
        ))}
        {isLoading && <ListLoading />}
      </div>

      {!isLoading && (
        <>
          {isError && <ListError />}
          {!isError && movies.length < 1 && <ListEmpty />}
          {hasMoreData && (
            <button
              onClick={onLoadMore}
              className="my-8 w-full text-center py-2 p-8 bg-gray-300 rounded-lg"
            >
              Load More
            </button>
          )}
        </>
      )}
    </>
  );
}

export default MovieList;
