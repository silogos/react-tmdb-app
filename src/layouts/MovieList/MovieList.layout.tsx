import { Outlet } from "react-router";

import { Filter, Hero } from "./MovieList.component";
import useMovieListLayout from "./MovieList.hook";

const categoryList = [
  {
    filterId: "/",
    title: "All",
  },
  {
    filterId: "/movies/now-playing",
    title: "Now Playing",
  },
  {
    filterId: "/movies/popular",
    title: "Popular",
  },
  {
    filterId: "/movies/top-rated",
    title: "Top Rated",
  },
  {
    filterId: "/movies/upcoming",
    title: "Upcoming",
  },
];

function MovieListLayout() {
  const { handleSearch } = useMovieListLayout();

  return (
    <>
      <div className="relative w-full min-h-screen">
        <Hero />

        <section id="movie-list" className="overflow-hidden py-4">
          <div className="mx-auto max-w-screen-xl px-6 lg:px-8">
            <Filter categoryList={categoryList} handleSearch={handleSearch} />
            <Outlet />
          </div>
        </section>
      </div>
    </>
  );
}

export default MovieListLayout;
