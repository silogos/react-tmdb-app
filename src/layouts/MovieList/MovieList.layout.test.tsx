import { describe, it, expect, vi, Mock } from "vitest";
import MovieList from "./MovieList.layout";
import { render } from "@testing-library/react";
import { BrowserRouter, useNavigate } from "react-router";
import userEvent from "@testing-library/user-event";

vi.mock("react-router", async () => {
  const originalModule = await vi.importActual("react-router");

  return {
    ...originalModule,
    useNavigate: vi.fn(),
  };
});

describe("MovieList Page", () => {
  it("renders correctly", () => {
    const component = render(
      <BrowserRouter>
        <MovieList />
      </BrowserRouter>
    );
    expect(component).toBeTruthy();
  });

  it("renders toMatchSnapshot", () => {
    const component = render(
      <BrowserRouter>
        <MovieList />
      </BrowserRouter>
    );
    expect(component).toMatchSnapshot();
  });

  it("search", async () => {
    const mockNavigate = vi.fn();
    (useNavigate as Mock).mockReturnValue(mockNavigate);

    const user = userEvent.setup();
    const { getByTestId } = render(
      <BrowserRouter>
        <MovieList />
      </BrowserRouter>
    );

    const inputSearch = getByTestId(
      "search-input"
    ) as unknown as HTMLInputElement;

    expect(inputSearch).toBeInTheDocument();

    inputSearch.focus();
    expect(inputSearch).toHaveFocus();

    await user.keyboard("hello{Enter}");

    expect(mockNavigate).toHaveBeenCalledWith(`/movies/search?query=hello`);
  });
});
