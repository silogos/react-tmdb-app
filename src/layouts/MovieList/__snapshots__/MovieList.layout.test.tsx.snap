// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieList Page > renders toMatchSnapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative w-full min-h-screen"
      >
        <section
          class="overflow-hidden py-4"
          id="hero"
        >
          <div
            class="mx-auto max-w-screen-xl px-4 py-14 lg:px-8 lg:py-20"
          >
            <h1
              class="text-center text-gray-300 text-4xl lg:text-7xl mb-4"
            >
              TMDB
            </h1>
            <p
              class="text-center text-gray-300 text-base lg:text-2xl"
            >
              (The Movie Database)
            </p>
          </div>
        </section>
        <section
          class="overflow-hidden py-4"
          id="movie-list"
        >
          <div
            class="mx-auto max-w-screen-xl px-6 lg:px-8"
          >
            <div
              class="flex flex-col lg:flex-row lg:items-center justify-between mb-4 gap-4"
            >
              <div
                class="flex flex-row flex-nowrap overflow-x-auto gap-4"
              >
                <a
                  aria-current="page"
                  class="px-4 py-2  text-nowrap rounded-lg bg-gray-50"
                  data-discover="true"
                  href="/"
                >
                  All
                </a>
                <a
                  class="px-4 py-2  text-nowrap rounded-lg bg-gray-400"
                  data-discover="true"
                  href="/movies/now-playing"
                >
                  Now Playing
                </a>
                <a
                  class="px-4 py-2  text-nowrap rounded-lg bg-gray-400"
                  data-discover="true"
                  href="/movies/popular"
                >
                  Popular
                </a>
                <a
                  class="px-4 py-2  text-nowrap rounded-lg bg-gray-400"
                  data-discover="true"
                  href="/movies/top-rated"
                >
                  Top Rated
                </a>
                <a
                  class="px-4 py-2  text-nowrap rounded-lg bg-gray-400"
                  data-discover="true"
                  href="/movies/upcoming"
                >
                  Upcoming
                </a>
              </div>
              <form
                novalidate=""
              >
                <input
                  class="w-full border-2 rounded-lg  px-4 py-2 bg-white"
                  data-testid="search-input"
                  name="query"
                  placeholder="Search..."
                  required=""
                  type="search"
                />
              </form>
            </div>
          </div>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative w-full min-h-screen"
    >
      <section
        class="overflow-hidden py-4"
        id="hero"
      >
        <div
          class="mx-auto max-w-screen-xl px-4 py-14 lg:px-8 lg:py-20"
        >
          <h1
            class="text-center text-gray-300 text-4xl lg:text-7xl mb-4"
          >
            TMDB
          </h1>
          <p
            class="text-center text-gray-300 text-base lg:text-2xl"
          >
            (The Movie Database)
          </p>
        </div>
      </section>
      <section
        class="overflow-hidden py-4"
        id="movie-list"
      >
        <div
          class="mx-auto max-w-screen-xl px-6 lg:px-8"
        >
          <div
            class="flex flex-col lg:flex-row lg:items-center justify-between mb-4 gap-4"
          >
            <div
              class="flex flex-row flex-nowrap overflow-x-auto gap-4"
            >
              <a
                aria-current="page"
                class="px-4 py-2  text-nowrap rounded-lg bg-gray-50"
                data-discover="true"
                href="/"
              >
                All
              </a>
              <a
                class="px-4 py-2  text-nowrap rounded-lg bg-gray-400"
                data-discover="true"
                href="/movies/now-playing"
              >
                Now Playing
              </a>
              <a
                class="px-4 py-2  text-nowrap rounded-lg bg-gray-400"
                data-discover="true"
                href="/movies/popular"
              >
                Popular
              </a>
              <a
                class="px-4 py-2  text-nowrap rounded-lg bg-gray-400"
                data-discover="true"
                href="/movies/top-rated"
              >
                Top Rated
              </a>
              <a
                class="px-4 py-2  text-nowrap rounded-lg bg-gray-400"
                data-discover="true"
                href="/movies/upcoming"
              >
                Upcoming
              </a>
            </div>
            <form
              novalidate=""
            >
              <input
                class="w-full border-2 rounded-lg  px-4 py-2 bg-white"
                data-testid="search-input"
                name="query"
                placeholder="Search..."
                required=""
                type="search"
              />
            </form>
          </div>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
